nova_vm/src/heap/element_array.rs

```diff
@@ -2304,6 +2373,18 @@ impl ElementArrays {
                 .contains(&Some(element)),
         }
     }
+
+    /// This method creates a "shallow clone" of the elements of a trivial/dense array.
+    /// It does not do anything with descriptors and assumes there is a previous validation in place.
+    pub fn shallow_clone(&mut self, elements_vector: ElementsVector) -> SealableElementsVector {
+        let new_len = usize::try_from(elements_vector.len()).unwrap();
+        let new_key = ElementArrayKey::from(new_len);
+        if new_key == elements_vector.cap {
```

Comment:
**issue:** As currently written, the comparison is the wrong way around: If the new key is the same as the old one, then we can use the new key. If it's different, then we must use the old key. **Note** that this ends up as just being "use the old key always".

Using the old key always is absolutely a valid choice. The downside is as I mentioned that it will overallocate when the source Array has a higher capacity than it would need nowadays. This is the `Array(256)` case I mentioned earlier.

Note that fixing that downside requires using an API different from `extend_from_within`; it would be an `extend_from_slice` from the source key+index to the `new_key` + new index (push + last_element_index).

But that's probably going to be too hard; you can just remove this if clause, and remove the `new_len`/`new_key` calculation. Just use the old key and old len directly.

Reply:
Oh, I see. Alright, I removed the conditional checks and hope at some point I can add support to that use case you mentioned.